### PR TITLE
 feat(sdk): Do not send a room subscription that has already been sent

### PR DESCRIPTION
**Pull request by Hywan**
This patch should ideally be reviewed commit-by-commit.

The problem the patch is trying to solve is the following: in sliding sync, for every new room subscriptions, all the previous room subscriptions are part of the request to be send. This is a waste of bandwidth. This patch adds the following feature: when a room subscription has been sent, it won't be part of the next request anymore. All room subscriptions are kept in memory though, but they are marked as `Applied` instead of `Pending` (see the new `RoomSubscriptionState` type).

---

* Address https://github.com/matrix-org/matrix-rust-sdk/issues/3647

### CODE REVIEW

**Comment by Hywan** [comment]
Night thought: If sticky parameters are "disabled" (i.e. the server doesn't return the `txn_id`), the new mechanism introduced by this PR will have no effect. All room subscriptions will continue being marked as `Pending`, and will be sent entirely with all requests. It does solve the problem with SS proxy, but not with SS native.

File: crates/matrix-sdk/src/sliding_sync/mod.rs [hunk 1]
```diff
@@ -928,17 +932,41 @@ impl SlidingSyncStickyParameters {
...
 
 impl StickyData for SlidingSyncStickyParameters {
     type Request = http::Request;
 
     fn apply(&self, request: &mut Self::Request) {
-        request.room_subscriptions = self.room_subscriptions.clone();
+        request.room_subscriptions = self
+            .room_subscriptions
+            .iter()
+            .filter_map(|(room_id, (state, room_subscription))| {
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/mod.rs:954]: In the constructor, it is a `map`, not a `filter_map`. Would you prefer the user to construct the room subscriptions with the default state every time? It feels annoying and error-prone to me.

File: crates/matrix-sdk/src/sliding_sync/sticky_parameters.rs [hunk 2]
```diff
@@ -49,6 +49,12 @@ pub trait StickyData {
 
     /// Apply the current data onto the request.
     fn apply(&self, request: &mut Self::Request);
+
+    /// Commit the current data, i.e. the request has been validated by a
+    /// response.
+    fn commit(&mut self) {
+        // noop
+    }
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/sticky_parameters.rs:57]: Yup I agree, good idea.

File: crates/matrix-sdk/src/sliding_sync/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bnjbvr) [reply crates/matrix-sdk/src/sliding_sync/mod.rs:954]: Ah you're right, we can keep it as is then!

**Pull request merged**